docs/Add-Models.md

```diff
@@ -0,0 +1,270 @@
+
+For example, in GPT2, we have the following breakdown:
+- `Gpt2Mlp`
+- `Gpt2Attention`
+- `Gpt2Block`: a block of Gpt2Mlp and Gpt2Attention
+- `Gpt2Transformer`: a stack of Gpt2Block
+- `Gpt2Embeddings`: token and position embeddings
+- `Gpt2LMHead`: a complete GPT2 model with embedding, transformer, and LM head.
+
+We follow the same breakdown in the implementation of Llama in Levanter.
+
+#### Note on the Implementation Format
+- Each class will have its key layers and components defined as attributes and be initialized with a static method `init()`.
+- Each class will inherit from `StateDictSerializationMixin` from `torch_serialization` and Equinox's `Module` class.
+
+### Mapping weights from/to Hugging Face
+To load weights from Hugging Face, you will need to write a class function `from_hf_state_dict()` in each of your model class. It takes in a Hugging Face state_dict and returns a Levanter state_dict.
+
+For implementation, there are a few helper classes from `torch_serialization` that you can use:
+- To add specific prefix to the keys of Hugging Face state_dict, you can use the helper function `apply_prefix()`. The prefix comes from the name of attributes defined at the beginning of your model class.
+- To unflatten the linear layers of Hugging Face, you can use the helper function `unflatten_linear_params()`.
+- To unstack the transformer blocks of Hugging Face, you can use the helper function `unstack_transformer_blocks()`.
+
+For example, below is the implementation of `from_hf_state_dict()` in LlamaAttention:
+
+```python
+def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None) -> "LlamaAttention":
+    # unflatten the linear layers of HF state_dict to match the shape of LlamaAttention
+    d = {}
+    d.update(unflatten_linear_layers(apply_prefix(prefix, "q_proj"), state_dict, self.q_proj, True))
+    d.update(unflatten_linear_layers(apply_prefix(prefix, "k_proj"), state_dict, self.k_proj, True))
+    d.update(unflatten_linear_layers(apply_prefix(prefix, "v_proj"), state_dict, self.v_proj, True))
+    d.update(unflatten_linear_layers(apply_prefix(prefix, "o_proj"), state_dict, self.o_proj, True))
+
+    return super().from_state_dict(d, prefix)
+```
+
+Similarly, to save weights to Hugging Face, you will need to write a class function `to_hf_state_dict()` in each of your model class.
+
+The correctness of your implementation can be validated through serialization tests, which will be discussed in the next section.
+
+## Write Tests
+There are two types of tests that you should write for your model: unit tests and serialization tests.
+
+### Unit Tests
+Unit tests are very useful for testing the correctness of your model implementation.
+It is recommended to have at least one test for each of your modules, so that you can make sure that each module is working as expected and capture any surprises early on, before you test them end-to-end.
```

Comment:
I honestly don't think this is necessary but it's useful to do if you get stuck to minimize.